unblob/handlers/compression/xz.py

```diff
@@ -1,110 +1,102 @@
 import io
-from typing import Optional, Tuple
+import lzma
+from typing import Optional
 
+import attr
+import hyperscan
 from structlog import get_logger
 
 from unblob.extractors import Command
 
-from ...file_utils import (
-    Endian,
-    convert_int8,
-    convert_int32,
-    decode_multibyte_integer,
-    iterate_patterns,
-    round_up,
-)
-from ...models import File, Handler, HexString, ValidChunk
+from ...file_utils import iterate_file
+from ...models import File, Handler, HexString, InvalidInputFormat, Regex, ValidChunk
 
 logger = get_logger()
 
 # The .xz file format definition: https://tukaani.org/xz/xz-file-format-1.0.4.txt
 
-XZ_PADDING = 4  # XZ format byte alignment
-MAX_MBI_LEN = 9  # maximum multi-byte integer size is 9, per XZ standard
-STREAM_HEADER_SIZE = 12  # magic bytes (6) + flags (2) + CRC32 (4)
-STREAM_FOOTER_SIZE = 12  # CRC32 (4) + backward size (4) + flags (2) + magic bytes (2)
-STREAM_FOOTER_MAGIC = b"YZ"
 
-MAGIC_BYTES_SIZE = 6
-BACKWARD_SIZE_LEN = 4
-CRC32_LEN = 4
-FLAG_LEN = 2
+STREAM_END_MAGIC_PATTERNS = [
+    Regex(r"\x00\x00YZ"),  # None
+    Regex(r"\x00\x01YZ"),  # CRC32
+    Regex(r"\x00\x04YZ"),  # CRC64
+    Regex(r"\x00\x0AYZ"),  # SHA-256
+]
```

Comment:
probably not, but was wondering if we should only match on flags that were present in the stream header? Or maybe at least check if the flag is the same on a match?

Reply:
I kept the same search rule so we don't generate a hyperscan db on every XZ magic match. I do however validate that concatenated streams have the same flags, tested by `test_xz_calculate_chunk_unique_stream` and `test_xz_calculate_chunk_unique_stream_padding`.